test/e2e/http_header_buffer_test.go

```diff
@@ -0,0 +1,310 @@
+				continue
+			}
+
+			for _, cond := range pod.Status.Conditions {
+				if cond.Type == corev1.PodReady && cond.Status == corev1.ConditionTrue {
+					return true, nil
+				}
+			}
+
+		}
+		t.Logf("waiting for new router pod for %s to become ready", ic.Name)
+		return false, nil
+	})
+
+	if pollErr != nil {
+		t.Errorf("timed out waiting for new router pod for %s to become ready: %v", ic.Name, pollErr)
+	}
+
+	name = name + "-fail-case"
+	clientPodInvalidRequest := buildCurlPod(name, echoRoute.Namespace, image, echoRoute.Spec.Host, service.Spec.ClusterIP, extraCurlArgs...)
+
+	if err := kclient.Create(context.TODO(), clientPodInvalidRequest); err != nil {
+		t.Fatalf("failed to create pod %s/%s: %v", clientPodInvalidRequest.Namespace, clientPodInvalidRequest.Name, err)
+	}
+
+	defer func() {
+		if err := kclient.Delete(context.TODO(), clientPodInvalidRequest); err != nil {
+			t.Fatalf("failed to delete pod %s/%s: %v", clientPodInvalidRequest.Namespace, clientPodInvalidRequest.Name, err)
+		}
+	}()
+
+	// Check curl pod logs for a 400 response since the sent headers
+	// are too large for HAProxy.
+	pollErr = wait.PollImmediate(1*time.Second, 3*time.Minute, func() (bool, error) {
+		readCloser, err := cl.CoreV1().Pods(clientPodInvalidRequest.Namespace).GetLogs(clientPodInvalidRequest.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).Stream(context.TODO())
+		if err != nil {
+			t.Logf("failed to read output from pod %s: %v", clientPodInvalidRequest.Name, err)
+			return false, nil
+		}
+		scanner := bufio.NewScanner(readCloser)
+		defer func() {
+			if err := readCloser.Close(); err != nil {
+				t.Errorf("failed to close reader for pod %s: %v", clientPodInvalidRequest.Name, err)
+			}
+		}()
+		for scanner.Scan() {
+			line := scanner.Text()
+			if strings.Contains(line, "400 Bad request") {
+				return true, nil
+			}
+		}
+		return false, nil
+	})
+	if pollErr != nil {
+		pod := &corev1.Pod{}
+		podName := types.NamespacedName{
+			Namespace: clientPodInvalidRequest.Namespace,
+			Name:      clientPodInvalidRequest.Name,
+		}
+		if err := kclient.Get(context.TODO(), podName, pod); err != nil {
+			t.Errorf("failed to get pod %s: %v", clientPodInvalidRequest.Name, err)
+		}
+
+		logs, err := cl.CoreV1().Pods(clientPodInvalidRequest.Namespace).GetLogs(clientPodInvalidRequest.Name, &corev1.PodLogOptions{
+			Container: "curl",
+			Follow:    false,
+		}).DoRaw(context.TODO())
+		if err != nil {
+			t.Errorf("failed to get logs from pod %s: %v", clientPodInvalidRequest.Name, err)
+		}
+
+		output := fmt.Sprintf("failed to observe the expected output: %v\nclient pod spec: %#v\nclient pod logs:\n%s", pollErr, pod, logs)
+		output = strings.Replace(output, random, fmt.Sprintf("<random data len %d btyes snipped>", length), 0)
+		t.Fatal(output)
+	}
+}
+
+// randomString quickly returns a string of
+// pseudo-random characters.
+func randomString(length int) (str string) {
+	b := make([]byte, length)
```

Comment:
So wasteful!
```suggestion
	b := make([]byte, (length+1)/2)
```
(Just kidding.  No need to micro-optimize this...)

Reply:
Alright I'll byte. How is this a micro-optimization? I must be missing something.

Reply:
You only need half the buffer space that the current logic allocates, and half the pseudo-random bytes that it generates.  The caller wants a string of the specified length, but `fmt.Sprintf("%x", ...)` produces 2 characters per byte, so the buffer only needs to be half as big as `length`, plus 1 to account for odd values for `length`.  What made me think of this was the `[:length]` subscript that you apply to the buffer, which didn't make sense to me until I realized `fmt.Sprintf` was producing twice as many characters as the caller requested.

Reply:
Fixed just because.